dsp/modules/tensorrt_llm.py

```diff
@@ -0,0 +1,229 @@
+        """Initialize the TensorRTModel.
+
+        Args:
+            model_name_or_path (str): The Huggingface ID or the path where tokenizer files exist.
+            engine_dir (str): The folder where the TensorRT .engine file exists.
+            **engine_kwargs (Optional[dict]): Additional engine loading keyword arguments.
+
+        Keyword Args:
+            use_py_session (bool, optional): Whether to use a Python session or not. Defaults to False.
+            lora_dir (str): The directory of LoRA adapter weights.
+            lora_task_uids (list[str]): list of LoRA task UIDs; use -1 to disable the LoRA module.
+            lora_ckpt_source (str): The source of the LoRA checkpoint.
+
+            If use_py_session is set to False, the following kwargs are supported:
+                max_batch_size (int, optional): The maximum batch size. Defaults to 1.
+                max_input_len (int, optional): The maximum input context length. Defaults to 1024.
+                max_output_len (int, optional): The maximum output context length. Defaults to 1024.
+                max_beam_width (int, optional): The maximum beam width, similar to `n` in OpenAI API. Defaults to 1.
+                max_attention_window_size (int, optional): The attention window size that controls the
+                    sliding window attention / cyclic KV cache behavior. Defaults to None.
+                sink_token_length (int, optional): The sink token length. Defaults to 1.
+        """
+        # Implementation here
+        self.model_name_or_path, self.engine_dir = model_name_or_path, engine_dir
+        super().__init__(model=self.model_name_or_path)
+        try:
+            import tensorrt_llm
+        except ImportError as exc:
+            raise ModuleNotFoundError(
+                "You need to install tensorrt-llm to use TensorRTModel",
+            ) from exc
+
+        try:
+            from transformers import AutoTokenizer
+        except ImportError as exc:
+            raise ModuleNotFoundError(
+                "You need to install torch and transformers",
+            ) from exc
+
+        # Configure tokenizer
+        self.tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name_or_path,
+            legacy=False,
+            padding_side="left",
+            truncation_side="left",
+            trust_remote_code=True,
+            use_fast=True,
+        )
+
+        self.pad_id = (
+            self.tokenizer.eos_token_id if self.tokenizer.pad_token_id is None else self.tokenizer.pad_token_id
+        )
+        self.end_id = self.tokenizer.eos_token_id
+
+        # Configure TensorRT
+        self.runtime_rank = tensorrt_llm.mpi_rank()
+        self.runner, self._runner_kwargs = load_tensorrt_model(engine_dir=self.engine_dir, **engine_kwargs)
+        self.history = []
+
+    def _generate(self, prompt: Union[list[dict[str, str]], str], **kwargs) -> list[str]:
```

Comment:
update return type for outputted response, kwargs dictionary. 
Ideally, self._generate should only return the response and all_kwargs can be updated separately. This can be done within the basic_request method as done for other LM providers

Reply:
I did some changes, not sure if I understood it right but let me know if this works. Thanks :)